Modules/_sqlite/connection.c

```diff
@@ -2270,13 +2326,59 @@ getlimit_impl(pysqlite_Connection *self, int category)
 }
 
 
+static PyObject *
+get_autocommit(pysqlite_Connection *self, void *Py_UNUSED(ctx))
+{
+    if (!pysqlite_check_thread(self) || !pysqlite_check_connection(self)) {
+        return NULL;
+    }
+    if (self->autocommit == AUTOCOMMIT_ENABLED) {
+        Py_RETURN_TRUE;
+    }
+    if (self->autocommit == AUTOCOMMIT_DISABLED) {
+        Py_RETURN_FALSE;
+    }
+    return PyLong_FromLong(DEPRECATED_TRANSACTION_CONTROL);
+}
+
+static int
+set_autocommit(pysqlite_Connection *self, PyObject *val, void *Py_UNUSED(ctx))
+{
+    if (!pysqlite_check_thread(self) || !pysqlite_check_connection(self)) {
+        return -1;
+    }
+    if (!autocommit_converter(val, &self->autocommit)) {
+        return -1;
+    }
+    if (self->autocommit == AUTOCOMMIT_ENABLED) {
+        if (!sqlite3_get_autocommit(self->db)) {
+            if (connection_exec_stmt(self, "COMMIT") < 0) {
+                return -1;
+            }
+        }
+    }
+    else if (self->autocommit == AUTOCOMMIT_DISABLED) {
+        if (!sqlite3_get_autocommit(self->db)) {
+            if (connection_exec_stmt(self, "COMMIT") < 0) {
+                return -1;
+            }
+        }
+        if (connection_exec_stmt(self, "BEGIN") < 0) {
+            return -1;
+        }
```

Comment:
@malemburg: I would like to hear your thoughts about this scenario. If `autocommit` is being set to `False` (aka `AUTOCOMMIT_DISABLED`) _and_ there is a pending transaction, we've got two options:

1. do nothing; after all `autocommit = False` implies an open transaction, so since we've already got one, there's nothing to do
2. commit the pending transaction and open a new one

As you can see, the PR follows the latter path: commit the pending transaction and open a new one. The changes you've proposed for PEP 249 does not seem to adress this case. Perhaps it would be best that PEP 249 define the semantics of `autocommit = False` when there already is an open transaction.

Reply:
I tried to cover this in https://discuss.python.org/t/introducing-a-pep-249-compliant-transaction-behaviour-to-sqlite3/16516/6?u=malemburg:
> The details around changing the value are database dependent, e.g. changing from True to False will usually start a new transaction, while changing from False to True will usually cause the current transaction to be committed. However, it is also possible that a database raises an error in case there are pending changes to be committed. To prevent the latter case, it is best to explicitly commit or rollback the current transaction before switching on autocommit mode on the connection.

When autocommit is `True`, there should not be any pending transactions, since all operations are automatically committed once executed. Setting autocommit to `False `will then turn off autocommit and open a new transaction.

When autocommit is `False `and there are pending transactions, the situation is less clear when switching to `True`. I'll still have to discuss this on the DB-SIG, since there are two possible options:

1. The database module raises an exception, since it's not clear what to do.
2. The module automatically commits the pending transaction, since that's what autcommit is all about.

Reply:
My preference would be option 2, as mentioned in https://discuss.python.org/t/introducing-a-pep-249-compliant-transaction-behaviour-to-sqlite3/16516/8?u=malemburg

Reply:
Rereading your remark, perhaps I misunderstood what you were asking. Are you asking what happens when `.autocommit` is set to `False`, if it already is `False `?

In that case, nothing should happen, since the status is not changed. Having this cause an implicit commit would be surprising.

Reply:
> Rereading your remark, perhaps I misunderstood what you were asking. Are you asking what happens when `.autocommit` is set to `False`, if it already is `False `?

Sorry, I see I formulated myself unclear. Consider this scenario:

- `con = sqlite3.connect(..., autocommit=True)`
- `con.execute("begin; select * from nosuchtable; commit")` <= this fails half way through
- `con.autocommit = False`

After thinking about, I agree that emitting an implicit commit is surprising. After all, what I'm trying to fix here is getting rid of the existing implicit transaction control, so introducing more implicit transaction control is of course counter-productive :)

BTW, I'll reply to your thread on the db-sig list.